updated inv_Phi

spinkney opened the pull request:
## Summary

faster inverse phi see issue #2555 
based on Algorithm described in Wichura, M. J. (1988) Algorithm AS 241: The percentage points of the normal distribution. Applied Statistics, 37, 477–484.
which provides precise results up to about 16 digits.

## Tests

I used the same tests that were already there

## Side Effects

None

## Release notes

Updated algorithm for inv_Phi which is approximately 2x faster with precision of 16 digits. Based on the Fortran algorithm described in Wichura, M. J. (1988) Algorithm AS 241: The percentage points of the normal distribution. Applied Statistics, 37, 477–484.

## Checklist

- [x] Math issue #2555 

- [x] Copyright holder: Sean Pinkney

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

spinkney commented:
Multiplying by a big int preserves accuracy.

```c++
  static long int ep = 10000000000;

  double new_p;
  int sign;
  if (p >= 0.9999) {
    new_p = (ep - ep * p) / ep;
    sign = -1;
  } else {
    new_p = p;
    sign = 1;
  }
```
the error is
```c++
p, exact CDF inverse, computed CDF inverse old, diff
...
0.999, 3.090232306167814, 3.090232306167823, -8.881784197001252e-15
0.99999, 4.264890793922825, 4.264890793923692, -8.668621376273222e-13
0.9999999000000001, 5.199337582192816, 5.199337582346428, -1.536113458655564e-10

Maximum error: 1.536113458655564e-10

p, exact CDF inverse, computed CDF inverse new, diff
...
0.999, 3.090232306167814, 3.090232306167813, 4.440892098500626e-16
0.99999, 4.264890793922825, 4.264890793922826, -8.881784197001252e-16
0.9999999000000001, 5.199337582192816, 5.199337582192817, -8.881784197001252e-16

Maximum error: 1.110223024625157e-15

p, exact CDF inverse, computed CDF inverse boost, diff
...
0.999, 3.090232306167814, 3.090232306167814, 0
0.99999, 4.264890793922825, 4.264890793923841, -1.016076112136943e-12
0.9999999000000001, 5.199337582192816, 5.199337582290661, -9.78443992494249e-11

Maximum error: 9.78443992494249e-11

Average error_new: 3.833738881908744e-16
Average error_old: 9.655581889589371e-12
Average error_boost: 6.178892467123553e-12

Is Error Better? True

error_new < error_boost < error_old
```

spinkney commented:
The speedups I'm getting vs base are
![plot_base](https://user-images.githubusercontent.com/23704057/131228584-88e9cd63-a243-4907-9b6a-9fbb2ef70a2b.png)

the speedups I'm getting vs boost are
![plot_boost](https://user-images.githubusercontent.com/23704057/131228589-17fa9a75-2e99-46d1-8b68-17c94656f857.png)

bob-carpenter commented:
Thanks for all the eval.  Looks great.  Just let me know when it's ready for final review.

spinkney commented:
@bob-carpenter it's ready. 
overview of updates:

- added a large int constant to the constants file
- inv_Phi calcs are all done in a lambda function. This is passed to the inv_Phi function to account for floating point error using the big int. 
- added some more tests

bob-carpenter commented on stan/math/prim/fun/inv_Phi.hpp:
```diff
@@ -37,45 +40,112 @@ inline double inv_Phi(double p) {
+/**
+ * The inverse of the unit normal cumulative distribution function.
+ *
+ * The ternary operation is to protect against floating point error
```
The implementation details don't belong in the doc.  It would be OK to talk about precision.

bob-carpenter commented on stan/math/prim/fun/inv_Phi.hpp:
```diff
@@ -37,45 +40,112 @@ inline double inv_Phi(double p) {
 
-  return x;
+/**
+ * The inverse of the unit normal cumulative distribution function.
```
This should be phrased in the form of the value it returns.  For example, "Return the value of the inverse standard normal cdf at the specified argument."

bob-carpenter commented on stan/math/prim/fun/inv_Phi.hpp:
```diff
@@ -37,45 +40,112 @@ inline double inv_Phi(double p) {
+ * in values near 1.
+ *
+ * @param p argument between 0 and 1 inclusive
+ * @return Real value of the inverse cdf for the standard normal distribution.
```
Only sentences should be capitalized and ended with a period.

bob-carpenter commented on test/unit/math/prim/fun/inv_Phi_test.cpp:
```diff
@@ -19,6 +19,39 @@ TEST(MathFunctions, inv_Phi) {
+         0.55,      0.65,    0.75,  0.85, 0.95, 0.999, 0.99999, 0.9999999};
+
+  double exact[]
+      = {-5.199337582192816931587347266962336866509737160238716454318950531,
```
This won't hurt anything, but they can also be cut down to 18 digits of precision as that's already more than you get with `double`.

bob-carpenter commented on test/unit/math/prim/fun/inv_Phi_test.cpp:
```diff
@@ -19,6 +19,39 @@ TEST(MathFunctions, inv_Phi) {
+  int numValues = sizeof(p) / sizeof(double);
+
+  for (int i = 0; i < numValues; ++i) {
+    EXPECT_NEAR(exact[i], inv_Phi(p[i]), 1.5e-15);
```
[comment]
This is really good precision!  You could mention this in the doc that it's 14 digits of precision for values in (0.0000001, 0.9999999).

What happens outside of the range you consider?

bob-carpenter commented on stan/math/prim/fun/inv_Phi.hpp:
```diff
@@ -21,13 +22,15 @@ namespace math {
  *
- * Algorithm first derived in 2003 by Peter Jon Aklam at
- * http://home.online.no/~pjacklam/notes/invnorm/
+ * Algorithm described in Wichura, M. J. (1988)
```
I'd like to see the algorithm reference up in the public-facing (non-internal) doc.  The internal function can have minimal doc.  The definition of Phi isn't necessary, but if it's included, it should be pointed out that the function is computing its inverse.

bob-carpenter commented on stan/math/prim/fun/constants.hpp:
```diff
@@ -26,6 +26,12 @@ inline double e() { return boost::math::constants::e<double>(); }
 inline double pi() { return boost::math::constants::pi<double>(); }
 
+/**
+ * The value of 10000000000, ie a big integer.
```
This constant should be local to the function that uses it.  Or at the very least put in the internal namespace.

If it does get documented somewhere, "ie" should be "i.e.," with two periods and a comma.  It should also mention why this value was chosen.

spinkney replied:
I've removed this constant from the constants file and put it in the internal namespace.

Interestingly, using INT_MAX fails the floating point accuracy. I did a loop from 1e10 to INT_MAX to find the biggest INT that doesn't fail. It's 2e10. That is added to the documentation of the functions.

bob-carpenter reviewed (changes requested):
The code and evals look great.  Thanks so much for doing this.

I'd have just accepted the doc as is if it weren't for wanting the magic number to not be in our top-level namespace.  So you can treat all the doc stuff as optional.  I included some high-level guidance to help going forward, even if you don't change this PR.  Also, some of the changes I requested weren't even in the code you touched, but can be easily fixed.  Thanks for bearing with me here.

spinkney commented:
@bob-carpenter updates are in and I created a docs PR at https://github.com/stan-dev/docs/pull/393.

bob-carpenter reviewed (approved):
Looks good.  Thanks for updating.
